### PR TITLE
First iteration on a new filter system to allow more granular configuration

### PR DESCRIPTION
**Pull request by oscarbc96**
### Improvements
- Add new `RuleConfig`, allows to overwrite the default behaviour of the rule changing rule mode and risk value.
- Add new `Filter`, allows setting custom rule configuration to matching coincidences.
- New RuleModes supported: `RuleMode.DISABLED` and `RuleMode.WHITELISTED`.
### Breaking changes
- Class variables `Rule.RULE_MODE` and `Rule.RISK_VALUE` should be changed to use properties `rule_mode` and `risk_value`. These properties take in consideration the custom config that might be applied.
- If rule mode is `DISABLED` or `WHITELISTED`; methods `add_failure_to_result` and `add_warning_to_result` will have no effect.
- `add_failure_to_result` and `add_warning_to_result` accepts a new optional parameter named `context`. This variable is going to be evaluated by filters defined in the custom config.

### CODE REVIEW

File: docs/rule_config.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+Allows to overwrite the default behaviour of the rule changing rule mode and risk value. It accepts a more granular
```
Comment (ocrawford555) [inline docs/rule_config.md]: ... of the rule, such as changing the rule mode and risk value. ...

File: docs/rule_config.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
+Allows to overwrite the default behaviour of the rule changing rule mode and risk value. It accepts a more granular 
+configuration using the filter.
+ 
+{{ inline_source('cfripper.config.rule_config.RuleConfig') }}
+
+## Filters
+
+Allows to set custom rule configuration to matching coincidences. When adding a failure it will check if there is a
```
Comment (ocrawford555) [inline docs/rule_config.md]: I would remove the first sentence probably, just start with `When adding a failure`

Also, it is for failure and warning I think

File: docs/rule_config.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
+Allows to overwrite the default behaviour of the rule changing rule mode and risk value. It accepts a more granular 
+configuration using the filter.
+ 
+{{ inline_source('cfripper.config.rule_config.RuleConfig') }}
+
+## Filters
+
+Allows to set custom rule configuration to matching coincidences. When adding a failure it will check if there is a 
+filter that match the current context and set the new risk or mode. Context depends on each rule and is available inside
```
Comment (ocrawford555) [inline docs/rule_config.md]: ... that matches the current context ...

File: docs/rule_config.md [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+| `not`      | Same as not a                                                               | `{"not": True}`                         |
+| `or`       | True if any arg is True                                                     | `{"or": [False, True]}`                 |
+| `and`      | True if all args are True                                                   | `{"and": [True, True]}`                 |
+| `in`       | Same as a in b                                                              | `{"in": ["b", ["a", "b"]]}`             |
+| `regex`    | True if b match pattern a                                                   | `{"regex": [r"^\d+$", "5"]}`            |
+| `ref`      | Get the value at any depth of the context based on the path described by a. | `{"ref": "param_a.param_b"}`            |
+
+### Examples
+
+Disable the rule if the role name is prefixed with `sandbox-` and the principal equals to `arn:aws:iam::123456789012:role/test-role`.
```
Comment (ocrawford555) [inline docs/rule_config.md]: `principal equals arn:aws...` (remove the 'to')

File: cfripper/rules/base_rules.py [hunk 5]
```diff
@@ -46,16 +59,24 @@ def add_failure_to_result(
...
-            resource_ids=resource_ids,
-            actions=actions,
-            granularity=granularity or self.GRANULARITY,
-        )
+        rule_mode = rule_mode or self.rule_mode
+        risk_value = risk_value or self.risk_value
+        for filter in self.rule_config.filters:
+            if filter(**context):
+                risk_value = filter.risk_value or risk_value
+                rule_mode = filter.rule_mode or rule_mode
```
[the remark this answers is not in the document]
Comment (jsoucheiron) [reply cfripper/rules/base_rules.py]: that's a very good point. We need to document filter precedence.

File: cfripper/rules/base_rules.py [hunk 6]
```diff
@@ -66,17 +87,25 @@ def add_warning_to_result(
...
-            resource_ids=resource_ids,
-            actions=actions,
-            risk_value=risk_value or self.RISK_VALUE,
-            rule_mode=rule_mode or self.RULE_MODE,
-        )
-        result.add_warning(warning)
+        rule_mode = rule_mode or self.rule_mode
+        risk_value = risk_value or self.risk_value
+        for filter in self.rule_config.filters:
+            if filter(**context):
```
Comment (jsoucheiron) [inline cfripper/rules/base_rules.py]: filter is a keyword :(

File: cfripper/rules/base_rules.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (oscarbc96) [reply cfripper/rules/base_rules.py]: Yes, this is in purpose. You will be able to put a first rule with a wide filter and for a subset change the behaviour using a second filter. Following the cascade, Taking preference always the last value set.

**Pull request merged**